### PR TITLE
Batch upload

### PR DESCRIPTION
**Pull request by b2mdevelopment**
this PR regarding story [106027294!](https://www.pivotaltracker.com/story/show/106027294)
It allow reyna to send messages in batches based on configuration provided by the caller
Caller have to set reyna mode is batch enabled/disabled. default is true
Caller have to set batch URL as this can be different per each caller
if batch URL not provider reyna will add /batch to the last message url
it will restrict the batch size to 100 messages or 300K
Reyna sleep 1 second per batch

there is another enhancement included where if post for host failed we don't sleep. instead we record that last request failed and exit immediately then next post will check if it still in blackout mode or not.

### CODE REVIEW

File: src/reyna/MessageProvider/BatchProvider.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            long maxBatchSize = this.BatchConfiguration.BatchMessagesSize;
+            while (message != null && count < maxMessagesCount && size < maxBatchSize)
+            {
+                headers = message.Headers;
+                batch.Add(message);
+                
+                size = this.GetSize(batch);
+                count++;
+                message = this.Repository.GetNextMessageAfter(message.Id);
+            }
```
Comment (aflugge) [inline src/reyna/MessageProvider/BatchProvider.cs:58]: In Android reyna you checked after the while if size was above maxBatchSize and in that case removed the last message, you are not doing that here, which means batches can be slightly larger than maxBatchSize, right?

File: src/reyna/Preferences/Preferences.cs [hunk 2]
```diff
@@ -83,6 +86,37 @@ public bool OffChargeBlackout
...
+            {
+                return GetRegistryValue(BatchUploadKeyName, true);
+            }
+        }
+
+        public Uri BatchUploadUrl
+        {
+            get
+            {
+                var url = GetRegistryValue(BatchUploadUriKeyName, string.Empty);
```
Comment (aflugge) [inline src/reyna/Preferences/Preferences.cs:101]: Just saw that we have that on Android as well, but why do we need to be able to set the URL for batch upload to anything but the default? Is that already for people using some sort of proxy?

File: src/reyna/MessageProvider/BatchProvider.cs (hunk 1 above)
Comment (b2mdevelopment) [inline src/reyna/MessageProvider/BatchProvider.cs]: same logic in android.

we check during adding the message to the batch.

if batch size exceeds we remove the last message

**Comment by b2mdevelopment** [comment]
as others, I don't understand whether we ever would need non-default url for batch uploading. Other than this, I'm :+1: this PR!
@ivanbokii

File: src/reyna/MessageProvider/BatchProvider.cs (hunk 1 above)
Comment (b2mdevelopment) [inline src/reyna/MessageProvider/BatchProvider.cs:58]: Handled. thanks Anton

**Pull request merged**